Storage and VM snapshots (#12)
EGI-Foundation/fedcloud-documentation: source/storage.rst

+--------------------+----------------------------------------+--------------+----------------------+---------------------------------+
| **Object Storage** | from any device connected to the       | Possible     | only for the data    | via HTTP requests to server     |
|                    | internet                               |              | stored               |                                 |
+--------------------+----------------------------------------+--------------+----------------------+---------------------------------+

According to your application needs, you may select one technology over the
other. In general, block storage is a good and simple solution for temporary
data and data which you do not need to share beside the single application
running on a single VM. If you need to have your data exposed within portals
or shared between different steps of your processing workflow, it is usually
best to use the object storage.

Block Storage
-------------

Block storage is created and managed via requests to specific APIs. Once the
storage is attached to a VM, it is managed as a regular disk that can be

> use as any other disk from within the VM.

Reviewer: ```suggestion
used as any other disk from within the VM.
```

+--------------------------------------------+--------------------------------------+
| ``volume delete <volume>``                 | deletes a volume                     |
+--------------------------------------------+--------------------------------------+
| ``server add volume <server> <volume>``    | Attach a volume to a server          |
+--------------------------------------------+--------------------------------------+
| ``server remove volume <server> <volume>`` | Dettach a volume from a server       |
+--------------------------------------------+--------------------------------------+

For using the legacy OCCI interface refer to `OCCI How-to <https://wiki.egi.eu/wiki/HOWTO11_How_to_use_the_rOCCI_Client>`_.

Using block storage from your VMs
:::::::::::::::::::::::::::::::::

Block Storage will appear as block devices into your VM. Usually these devices
are empty upon creation. The first time you attach them to a VM, you will need
to partition and create filesystems on the device.

You can just create a filesystem on the block device with the following
command (run this at your VM!).

.. warning::

>    Only run this command the first time you use the device, it removes all data storedi

Reviewer: Missing word or non wanted last letter?

Author: not wanted last letter, fixed

.. code-block:: console

   # mkfs.ext4 /dev/<volume device>

The ``<volume device>`` (e.g ``vdb``) can be obtained with the ``openstack volume show``
command

Once you have a filesystem you can mount it at the desired path:

.. code-block:: console

   # mount /dev/<volume device> /<path>

With that you can access ``/<path>`` where all your data will be available.
Applications will not see any difference between a block storage device and a

> regular hardware disk, thus no major changes should be required in the application logic.

Reviewer: We are speaking about VMs, maybe we can drop the `hardware` term.

Author: :+1:

Note that some OS, like `CERNVM <https://appdb.egi.eu/store/vappliance/cernvm/ CERNVM>`_,

> will automatically detect all the attached block storage and add it to theo

Reviewer: ```suggestion
will automatically detect all the attached block storage and mount it under the
```
Not sure about the intended meaning, it may seem a bit obscure.

Author: removed the whole paragraph, not really bringing much

.. toctree::

Storage on EGI Cloud
====================

This page aims to give a brief description of the storage features provided by
the EGI Cloud and a basic tutorial on how to use and integrate them into your
application.

If you are in need of more storage than the one provided within image disk of
your VMs, or need to persist data independently from the VMs storage, you will

> need to use one of the solutions available on the EGI service catalogue.

Reviewer: ```suggestion
need to use one of the solutions available on the `EGI service catalogue <https://www.egi.eu/services/>`_.
```

their own set of advantages and disadvantages

**Block storage** provides additional storage blocks which can be attached to a
virtual machine. A storage block is a virtual disk of a given size, which can 
be exposed as a virtual device in the VM. You can think of this can of devices
as a USB stick that can be plugged into the VMs and can be used as a regular
drive. You can format it with any file system you want and mount it in your VM.
Block devices are persistent, thus they keep all the data after VM shutdown and
need to be explicitly destroyed when data is not needed anymore. Block storage
disks can be accessed only from within a VM, and only from VMs running at the
same provider where the block storage is located. Also, they can be accessed by
only one VM at the same time. As part of the IaaS service, block storage is
managed via OpenStack/OCCI APIs. There is a limit on the number of block
storage devices you can attach on a VM and there is a limit to the maximum
size of such virtual disks. These values will depend on the particular provider
and your SLA. Moreover, the disk space is accounted for the entire block
storage device, regardless how much of it is actually used. 

> **Object storage** is a standalone service that stores data as set of individual

Reviewer: ```suggestion
**Object storage** is a standalone service that stores data as sets of individual
```

Either use `a set of` or `sets of`

> objects, which can have different types (e.g. files, images, documents) and

Reviewer: Do objects in swift really have a type? Or is the intent to say that objects can be of any/various file types?

Reviewer: I don't see much things here: https://docs.openstack.org/api-ref/object-store/index.html?expanded=get-object-content-and-metadata-detail#objects

Author: the intent is to say objects can be of whatever type, I'm rewording

are organized within containers (e.g. folders). Each object/file has is own
URL, which can be used to access the resource, share the file with other
people, setup custom metadata and access control lists. These objects are

> accessed and managed via a REST API (OpenStack SWIFT). Differently from the

Reviewer: ```suggestion
accessed and managed via a REST API (OpenStack `SWIFT <https://docs.openstack.org/api-ref/object-store/index.html>`_). Differently from the
```

|                    | site the VM is located                 |              |                      |                                 |
+--------------------+----------------------------------------+--------------+----------------------+---------------------------------+
| **Object Storage** | from any device connected to the       | Possible     | only for the data    | via HTTP requests to server     |
|                    | internet                               |              | stored               |                                 |
+--------------------+----------------------------------------+--------------+----------------------+---------------------------------+

According to your application needs, you may select one technology over the
other. In general, block storage is a good and simple solution for temporary
data and data which you do not need to share beside the single application
running on a single VM. If you need to have your data exposed within portals
or shared between different steps of your processing workflow, it is usually
best to use the object storage.

> 

Reviewer: ```suggestion
```

Drop duplicated empty line

Block Storage
-------------

Block storage is created and managed via requests to specific APIs. Once the

> storage is attached to a VM, is managed as a regular block device that can be

Reviewer: ```suggestion
storage is attached to a VM, it is managed as a regular block device that can be
```

Reviewer: I think that the term `block device` will probably be confusing/unclear lots of people

managed from within the VM as any other block device.

Management
::::::::::

OpenStack providers offer block storage using their native API. Both the 
CLI client or dashboard can be used. Main commands for managing volumes are
listed below:

+------------------------------------------+--------------------------------------+
| Command                                  | function                             |
+==========================================+======================================+

> | `volume create --size <size> <name>`     | create a volume of size `<size>` GBs |

Reviewer: You may want to use double backquotes here

Author: I do...

|                                          | and name `<name>`                    |
+------------------------------------------+--------------------------------------+
| `volume list`                            | list available volumes               | 
+------------------------------------------+--------------------------------------+
| `volume show <volume>`                   | show details of a given volume       | 
+------------------------------------------+--------------------------------------+
| `volume delete <volume>`                 | deletes a volume                     | 
+------------------------------------------+--------------------------------------+
| `server add volume <server> <volume>`    | Attach a volume to a server          |
+------------------------------------------+--------------------------------------+
| `server remove volume <server> <volume>` | Dettach a volume from a server       |
+------------------------------------------+--------------------------------------+

> 

Reviewer: ```suggestion
```

> For using the legacy OCCI interface refer to `OCCI Howto <https://wiki.egi.eu/wiki/HOWTO11_How_to_use_the_rOCCI_Client>`_.

Reviewer: ```suggestion
For using the legacy OCCI interface refer to `OCCI How-to <https://wiki.egi.eu/wiki/HOWTO11_How_to_use_the_rOCCI_Client>`_.
```

Using block storage from your VMs
:::::::::::::::::::::::::::::::::

Block Storage will appear as block devices into your VM. Usually these devices
are empty upon creation. The first time you attach them to a VM, you will need
to partition and create filesystems on the device.

You can just create a filesystem on the block device with the following
command (run this at your VM!).

.. warning:: 

   Only run this command the first time you use the device, it removes all data storedi
   
.. code-block:: console

   # mkfs.ext4 /dev/<volume device>

> The `<volume device>` (e.g `vdb`) can be obtained with the `openstack volume show`

Reviewer: ```suggestion
The ``<volume device>`` (e.g ``vdb``) can be obtained with the ``openstack volume show``
```

command

Once you have a filesystem you can mount it at the desired path:

.. code-block:: console

   # mount /dev/<volume device> /<path>

> With that you can access `/<path>` where all your data will be available.

Reviewer: ```suggestion
With that you can access ``/<path>`` where all your data will be available.
```

Applications will not see any difference between a block storage device and a
regular hardware disk, thus no major changes should be required in the application logic.

> Note that some OS, like [https://appdb.egi.eu/store/vappliance/cernvm/ CERNVM],

Reviewer: ```suggestion
Note that some OS, like `CERNVM <https://appdb.egi.eu/store/vappliance/cernvm/ CERNVM>`_,
```

will automatically detect all the attached block storage and add it to theo
root virtual file system.

Object Storage
--------------

.. note::

   We are currently re-evaluating the object storage capabilities of EGI
   Cloud and will update this documentation as soon as possible.

> 

Reviewer: ```suggestion
```

OpenStack SWIFT offers a RESTful API to manage your storage. As with block
storage you can manage it via the OpenStack CLI or web dashboard.

Available SWIFT providers resources can be discovered in `GOCDB <https://goc.egi.eu/portal/index.php?Page_Type=Services&serviceType=org.openstack.swift&selectItemserviceType=org.openstack.swift&ngi=&searchTerm=&production=TRUE&monitored=TRUE&certStatus=Certified&scopeMatch=all&servKeyNames=&servKeyValue=>`_.

> For accessing the endpoint check the `URL` of the specific provider.

Reviewer: ```suggestion
For accessing the endpoint check the ``URL`` of the specific provider.
```
